Refactor URLPattern's canonical and tokenizers modules to preemptively support URLPattern parser module

jelee53 opened the pull request:
#### 65cb4e7a01154f36a4f3ed69ac0b08c55e892480
<pre>
Refactor URLPattern&apos;s canonical and tokenizers modules to preemptively support URLPattern parser module
<a href="https://bugs.webkit.org/show_bug.cgi?id=282418">https://bugs.webkit.org/show_bug.cgi?id=282418</a>
<a href="https://rdar.apple.com/139054089">rdar://139054089</a>

Reviewed by Sihui Liu and Chris Dumez.

Both URLPatternCanonical.cpp and URLPatternTokenizer.cpp both need to be updated to support the incoming URLPatternParser changes.

* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizePath):
(WebCore::canonicalizePathWrapper):
(WebCore::callEncodingCallback):
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.h:
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
(WebCore::URLPatternUtilities::Token::isNull const):
(WebCore::URLPatternUtilities::Token::isInitialized const):
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:

Canonical link: <a href="https://commits.webkit.org/286066@main">https://commits.webkit.org/286066@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65ea868711693283fb6ab531e4b6a2a2678385ad

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74566 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53995 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27377 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78998 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25804 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76683 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63128 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1780 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58608 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16887 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77633 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48744 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64112 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39004 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45815 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21608 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24137 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67155 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80476 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1883 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1109 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66857 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2031 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64130 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66144 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16448 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10094 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8248 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1848 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4635 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1876 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/2797 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1883 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -185,30 +185,35 @@ static ExceptionOr<String> canonicalizeOpaquePath(const String& value)
 
 // https://urlpattern.spec.whatwg.org/#canonicalize-a-pathname, combined with https://urlpattern.spec.whatwg.org/#process-pathname-for-init
-ExceptionOr<String> canonicalizePath(const String& pathnameValue, StringView protocolValue, BaseURLStringType pathnameValueType)
+static ExceptionOr<String> canonicalizePath(const String& pathnameValue)
```
We could name this to "canonicalizePathname" as spec

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -185,30 +185,35 @@ static ExceptionOr<String> canonicalizeOpaquePath(const String& value)
+    return result.toString();
+}
+
+ExceptionOr<String> canonicalizePathWrapper(const String& pathnameValue, StringView protocolValue, BaseURLStringType pathnameValueType)
```
And rename this to "processPathname"

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -185,30 +185,35 @@ static ExceptionOr<String> canonicalizeOpaquePath(const String& value)
-ExceptionOr<String> canonicalizePath(const String& pathnameValue, StringView protocolValue, BaseURLStringType pathnameValueType)
+static ExceptionOr<String> canonicalizePath(const String& pathnameValue)
+{
+    bool hasLeadingSlash = pathnameValue[0] == '/';
```
For safety, we need null check before accessing input like `pathnameValue[0]`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -252,4 +257,30 @@ ExceptionOr<String> canonicalizeHash(const String& value, BaseURLStringType valu
     return dummyURL.fragmentIdentifier().toString();
 }
 
+ExceptionOr<String> callEncodingCallback(EncodingCallbackType urlComponent, String&& input)
```
We normally make the parameter name match type, e.g. `EncodingCallbackType type`

jelee53 replied:
But if I do that, the style script complains that the parameter is not adding more information. 

Is there a way to bypass that complaint?

szewai replied:
Yes, you could just omit it in the declaration:
`ExceptionOr<String> callEncodingCallback(EncodingCallbackType, String&& input)`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -252,4 +257,30 @@ ExceptionOr<String> canonicalizeHash(const String& value, BaseURLStringType valu
+    case EncodingCallbackType::Hash:
+        return canonicalizeHash(input, BaseURLStringType::URL);
+    default:
+        return Exception { ExceptionCode::TypeError, "Invalid input to calling encoding callback."_s };
```
And this should be `Invalid input type for encoding callback`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -252,4 +257,30 @@ ExceptionOr<String> canonicalizeHash(const String& value, BaseURLStringType valu
     return dummyURL.fragmentIdentifier().toString();
 }
 
+ExceptionOr<String> callEncodingCallback(EncodingCallbackType urlComponent, String&& input)
```
Unclear why input is a `String&&` given that the implementation doesn't `WTFMove()` it at all.

jelee53 replied:
Oops my bad! Thank you for catching this!

I'll refactor so that invoking the canonicalize function will have WTFMove in the parameters as well.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -43,7 +43,12 @@ static bool isValidNameCodepoint(UChar codepoint, bool first)
 bool Token::isNull() const
 {
-    return type == TokenType::Open && !index && !value;
+    return !index;
```
Is it possible that index is null but value is not? If not, we probably want to assert here to avoid mistake:
```
if (!index) {
    ASSERT(value.isNull());
    return true;
}
return false;
```

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -43,7 +43,12 @@ static bool isValidNameCodepoint(UChar codepoint, bool first)
+    return !index;
+}
+
+bool Token::isInitialized() const
```
Could you point me to the corresponding description in spec? I cannot find it by searching "initialized"

jelee53 replied:
For me, I think it was a lot more readable to make a function isInitialized rather than doing !isNull(). 

But I can change it back!

szewai replied:
Got it; yeah I think we probably want to stick to `!isNull()` if spec says "if token is not null", since that will be easier for us to reason about changes the when we compare our implementation against spec.

szewai reviewed (approved):
LGTM with nits

jelee53 committed:
Refactor URLPattern's canonical and tokenizers modules to preemptively support URLPattern parser module
https://bugs.webkit.org/show_bug.cgi?id=282418
rdar://139054089

Reviewed by Sihui Liu and Chris Dumez.

Both URLPatternCanonical.cpp and URLPatternTokenizer.cpp both need to be updated to support the incoming URLPatternParser changes.

* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizePath):
(WebCore::canonicalizePathWrapper):
(WebCore::callEncodingCallback):
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.h:
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
(WebCore::URLPatternUtilities::Token::isNull const):
(WebCore::URLPatternUtilities::Token::isInitialized const):
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:

Canonical link: https://commits.webkit.org/286066@main
